jpeg.go

```diff
@@ -5,49 +5,156 @@ import (
 	"encoding/binary"
 	"errors"
 	"io"
+	"io/ioutil"
 )
 
 var headerBytes []byte = []byte{0xff, 0xd8}
 
+// scanReader reads until the scan terminating bytes are encountered
+type scanReader struct {
+	reader *io.Reader
+}
+
+func firstMarker(p []byte) int {
+	markerIndex := -1
+	// Look for an 0xff followed by anything other than 0x00 or [0xd0-0xd7] (those are RST markers)
+	for i := 0x01; i <= 0xff; i++ {
+		if i >= 0xd0 && i <= 0xd7 {
+			continue
+		}
+		markerIndex = bytes.Index(p, []byte{0xff, byte(i)})
+		if markerIndex >= 0 {
+			return markerIndex
+		}
+	}
+	return markerIndex
+}
+
+func (sr *scanReader) Read(p []byte) (n int, err error) {
+	buf := new(bytes.Buffer)
+	teedReader := io.TeeReader(*sr.reader, buf)
+	// kick the underlying reader into giving us more than just one byte because the
+	// the naive multireader will just give us one byte from the first reader in its arguments
+	if len(p) > 1 {
+		if n, err = io.ReadAtLeast(teedReader, p, 2); err != nil {
+			if errors.Is(err, io.EOF) {
+				return
+			} else if errors.Is(err, io.ErrUnexpectedEOF) {
+				// safe to just deal with on the next pass through
+				err = nil
+			} else {
+				err = &MalformedDataError{"unexpected error reading inside scan data", err}
+				return
+			}
+		}
+	} else {
+		// if len(p) is only 1, we can still attempt to read because we'll still
+		// peak the next byte and can act accordingly
+		if n, err = teedReader.Read(p); err != nil {
+			if errors.Is(err, io.EOF) {
+				return
+			}
+			err = &MalformedDataError{"unexpected error reading inside scan data", err}
+			return
+		}
+	}
+
+	// attempt to peek one more byte to ensure that the marker doesn't overlap
+	// the bytes returned
+	lastByte := make([]byte, 1)
+	var lastByteN int
+	if lastByteN, err = io.ReadFull(teedReader, lastByte); err != nil {
+		if errors.Is(err, io.EOF) || errors.Is(err, io.ErrUnexpectedEOF) {
+			// an EOF is ok here because we'll catch it normally on the next pass through
+			err = nil
+		} else {
+			err = &MalformedDataError{"unexpected error reading inside scan data", err}
+			return
+		}
+	}
+
+	var markerIndex int
+	if lastByteN == 1 {
+		markerIndex = firstMarker(append(p[:n], lastByte[0]))
+	} else {
+		markerIndex = firstMarker(p[:n])
+	}
+
+	if markerIndex >= 0 {
+		n = markerIndex
```

Comment:
I think this may need to be corrected in the case that `lastByteN > 0`.  That would mean that `markerIndex` may actually exceed what's been read into `p`.  It would be a pretty unlikely edge case, but still possible I think.

Reply:
Ah - yeah because `markerIndex` is the *beginning* of a 2 byte pattern and we've only added a single byte, `markerIndex` can't ever exceed the number of bytes read into `p`. I'm open to being wrong about this, but I _think_ it's correct as is?

Reply:
Oh yeah you're right, sorry!